File: 2048.cpp
#include <iostream>
#include <iomanip>
#include <cstdlib>
#include <ctime>
#include <cmath>
#include <cstring>
#include <ctype.h>
using namespace std;

struct node
{
    int pts;
    int **dat;
    node *prev;
    node *next;
};

class zy2048
{
    enum winState {playing,lose,win,both};
    friend ostream &operator<<(ostream &os,const zy2048 &obj);

private:
    const int order;
    int **data;
    int score;
    int cnt;
    winState state;
    node *head;
    node *r;
    void zeroTop(int c);

public:
    zy2048(int n);
    ~zy2048();
    int maxNum() const;
    void rev();
    void trn();
    void upward();
    void downward();
    void leftward();
    void rightward();
    void appear();
    bool efficient();
    void winlose();
    void record();
    void undo();
    void play();
};

zy2048::zy2048(int n):order(n),score(0),cnt(0),state(playing),head(new node),r(head)
{
    int i,j;
    data=new int*[n];
    for(i=0;i<n;i++)
        data[i]=new int[n];
    for(i=0;i<n;i++)
        for(j=0;j<n;j++)
            data[i][j]=0;

    srand(time(NULL));
    head->prev=NULL;
    record();
}

zy2048::~zy2048()
{
    int i;
    for(i=0;i<order;i++)
        delete [] data[i];
    delete [] data;
    node *p=head,*q=head;
    while(q)
    {
        q=q->next;
        delete p;
        p=q;
    }
}

int zy2048::maxNum() const
{
    int i,j;
    int max=0;
    for(i=0;i<order;i++)
        for(j=0;j<order;j++)
            max=data[i][j]>max ? data[i][j] : max;
    return max;
}

void zy2048::zeroTop(int c)
{
    int i,pos=0;
    int zeroNum=0;
    for(i=0;i<order;i++)
        if(data[c][i]==0)
            zeroNum++;
    for(i=0;i<order;i++)
        if(data[c][i]!=0)
        {
            data[c][pos]=data[c][i];
            pos++;
        }
    for(i=order-zeroNum;i<order;i++)
        data[c][i]=0;
}

void zy2048::rev()
{
    int i,j;
    int tmp;
    for(i=0;i<order;i++)
        for(j=0;j<order/2;j++)
        {
            tmp=data[i][j];
            data[i][j]=data[i][order-j-1];
            data[i][order-j-1]=tmp;
        }
}

void zy2048::trn()
{
    int i,j;
    int tmp;
    for(i=0;i<order;i++)
        for(j=i+1;j<order;j++)
        {
            tmp=data[i][j];
            data[i][j]=data[j][i];
            data[j][i]=tmp;
        }
}

void zy2048::upward()
{
    trn();
    leftward();
    trn();
}

void zy2048::downward()
{
    trn();
    rightward();
    trn();
}

void zy2048::leftward()
{
    int i,j;
    for(i=0;i<order;i++)
        for(j=0;j<order-1;j++)
        {
            zeroTop(i);
            if(data[i][j]==data[i][j+1])
            {
                data[i][j]=data[i][j]+data[i][j+1];
                data[i][j+1]=0;
                score+=data[i][j];
            }
        }
}

void zy2048::rightward()
{
    rev();
    leftward();
    rev();
}

void zy2048::appear()
{
    int appearNum;
    int x,y;
    appearNum=rand()%10==0 ? 4 : 2;
    do
    {
        x=rand()%order;
        y=rand()%order;
    }while(data[x][y]!=0);
    data[x][y]=appearNum;
}

void zy2048::winlose()
{
    int i,j;
    if(maxNum()>=2048)
        state=win;
    for(i=0;i<order;i++)
        for(j=0;j<order;j++)
            if(data[i][j]==0)
                return;
    for(i=0;i<order;i++)
        for(j=0;j<order-1;j++)
            if(data[i][j]==data[i][j+1] || data[j][i]==data[j+1][i])
                return;
    state=winState(state+lose);
}


bool zy2048::efficient()
{
    int i,j;
    for(i=0;i<order;i++)
        for(j=0;j<order;j++)
            if(r->dat[i][j]!=data[i][j])
                return true;
    return false;
}

void zy2048::record()
{
    int i,j;
    r->next=new node;
    r->next->prev=r;
    r=r->next;
    r->pts=score;
    r->dat=new int*[order];
    for(i=0;i<order;i++)
    {
        r->dat[i]=new int[order];
        for(j=0;j<order;j++)
            r->dat[i][j]=data[i][j];
    }
    r->next=NULL;
    cnt++;
}

void zy2048::undo()
{
    int i,j;
    node *p=r->prev;

    for(i=0;i<order;i++)
        delete []r->dat[i];
    delete r->dat;
    delete r;
    r=p;
    r->next=NULL;

    score=r->pts;
    for(i=0;i<order;i++)
        for(j=0;j<order;j++)
            data[i][j]=r->dat[i][j];
    cnt--;
}

ostream &operator<<(ostream &os,const zy2048 &obj)
{
    int i,j;
    int width=obj.maxNum()==0 ? 3 : log(obj.maxNum())/log(10)+3;
    for(i=0;i<obj.order;i++)
    {
        for(j=0;j<obj.order;j++)
            os<<left<<setw(width)<<obj.data[i][j]<<' ';
        os<<endl;
    }
    os<<"Your score: "<<obj.score<<endl;

    return os;
}
void zy2048::play()
{
    char dire[81];
    char d;
    appear();
    cnt=-1;

    while(true)
    {
        if(efficient())
        {
            appear();
            record();
        }
        winlose();
        cout<<endl<<"Step "<<cnt<<":"<<endl;
        cout<<*this<<endl<<endl;
        if(state==lose || state==both)
            break;
        cout<<"Please enter a direction(w/s/a/d) [Press z to undo, q to quit] :";
        cin.getline(dire,80,'\n');
        if(strlen(dire)>1)
        {
            cout<<"Wrong input!"<<endl;
            continue;
        }
        d=dire[0];

        switch(toupper(d))
        {
            case 'W': upward(); break;
            case 'S': downward(); break;
            case 'A': leftward(); break;
            case 'D': rightward(); break;
            case 'Z':
                if(r->prev!=head->next)
                    undo();
                else
                    cout<<"Cannot undo anymore!"<<endl;
                continue;
            case 'Q': break;
            default: cout<<"Wrong input!"<<endl; continue;
        }
        if(toupper(d)=='Q')
            break;

        if(!efficient())
            cout<<"This move is not efficient!"<<endl;
    }

    switch(state)
    {
        case lose: cout<<"You lose!"<<endl; break;
        case both: cout<<"You both win and lose!"<<endl;
    }
}

int main()
{
    zy2048 a(4);
    a.play();
    return 0;
}

File: 2048.py
# -*- coding: utf-8 -*-
import random
import copy

class Game:

    def __init__(self):
        while True:
            order=int(input("Please enter the order(2-100)\n"))
            if 1<order<101:
                break
        self.score=[0]
        self.points=0
        self.order=order
        self.record={}
        self.matrix=[[0 for i in xrange(self.order)] for i in xrange(self.order)]
        self.matrix=self.randBlock(self.matrix)
        self.matrix=self.randBlock(self.matrix)
        self.counter=0
        self.won=False
        self.record[0]=copy.deepcopy(self.matrix)
        self.outBlock(self.matrix)

    def zeroTop(self,lst):
        mat=copy.deepcopy(lst)
        zeroNum=mat.count(0)
        while mat.count(0)>0:
            mat.remove(0)
        return zeroNum*[0]+mat

    def combine(self,lst):
        mat=copy.deepcopy(lst)
        for i in range(len(mat)-1,0,-1):
            if mat[i]==mat[i-1]:
                mat[i]=mat[i]+mat[i-1]
                mat[i-1]=0
                self.points+=mat[i]
                mat=self.zeroTop(mat)
        return mat

    def rightward(self,matrix):
        mat=copy.deepcopy(matrix)
        for i in range(0,len(mat)):
            mat[i]=self.combine(self.zeroTop(mat[i]))
        return mat

    def leftward(self,matrix):
        mat=copy.deepcopy(matrix)
        for i in range(0,len(mat)):
            mat[i].reverse()
            mat[i]=self.combine(self.zeroTop(mat[i]))
            mat[i].reverse()
        return mat

    def trn(self,matrix):
        mat=copy.deepcopy(matrix)
        for i in range(0,len(mat)):
            for j in range(0,i):
                mat[i][j],mat[j][i]=mat[j][i],mat[i][j]
        return mat

    def downward(self,matrix):
        mat=copy.deepcopy(matrix)
        mat=self.trn(mat)
        mat=self.rightward(mat)
        mat=self.trn(mat)
        return mat

    def upward(self,matrix):
        mat=copy.deepcopy(matrix)
        mat=self.trn(mat)
        mat=self.leftward(mat)
        mat=self.trn(mat)
        return mat

    def moveBlock(self,matrix,direction):
        mat=copy.deepcopy(matrix)
        self.points=0
        if direction=="w":
            return self.upward(mat)
        elif direction=="s":
            return self.downward(mat)
        elif direction=="a":
            return self.leftward(mat)
        elif direction=="d":
            return self.rightward(mat)

    def randBlock(self,matrix):
        mat=copy.deepcopy(matrix)
        if random.random<0.1:
            appearNum=4
        else:
            appearNum=2
        length=len(mat)
        while True:
            row=random.randrange(0,length)
            col=random.randrange(0,length)
            if mat[row][col]==0:
                mat[row][col]=appearNum
                return mat

    def upAble(self,matrix):
        mat=copy.deepcopy(matrix)
        mat=self.upward(mat)
        if mat==matrix:
            return False
        else:
            return True

    def downAble(self,matrix):
        mat=copy.deepcopy(matrix)
        mat=self.downward(mat)
        if mat==matrix:
            return False
        else:
            return True

    def leftAble(self,matrix):
        mat=copy.deepcopy(matrix)
        mat=self.leftward(mat)
        if mat==matrix:
            return False
        else:
            return True

    def rightAble(self,matrix):
        mat=copy.deepcopy(matrix)
        mat=self.rightward(mat)
        if mat==matrix:
            return False
        else:
            return True

    def moveAble(self,matrix):
        mat=copy.deepcopy(matrix)
        if self.upAble(mat) or self.downAble(mat) or self.leftAble(mat) or self.rightAble(mat):
            return True
        else:
            return False

    def winBlock(self,matrix):
        mat=copy.deepcopy(matrix)
        temp=[]
        for i in range(0,len(mat)):
            temp=temp+mat[i]
        if 2048 in temp:
            return "win"
        elif self.moveAble(mat):
            return "playing"
        else:
            return "lose"

    def outBlock(self,matrix):
        for i in range(0,len(matrix)):
            print matrix[i]
        print "Your score:%d" % (self.score[self.counter])

    def play(self):
        while True:
            temp=copy.deepcopy(self.matrix)
            direction=raw_input("direction:w/s/a/d,z to undo\n>>>").lower()
            if direction in ["w","s","a","d"]:
                self.matrix=self.moveBlock(self.matrix,direction)
                if temp!=self.matrix:
                    self.matrix=self.randBlock(self.matrix)
                    self.counter+=1
                    self.record[self.counter]=copy.deepcopy(self.matrix)
                    self.score.append(self.score[-1]+self.points)
                else:
                    print "This move has no efficiency!"
            elif direction=="z":
                if self.counter>0:
                    self.counter-=1
                    self.matrix=copy.deepcopy(self.record[self.counter])
                    del self.record[self.counter]
                    del self.score[-1]
                else:
                    print "Cannot undo anymore!"
            else:
                print "Wrong input!"
            self.outBlock(self.matrix)
            if self.winBlock(self.matrix)=="win" and not(self.won):
                    print "You win!"
                    self.won=True
            if self.winBlock(self.matrix)=="lose":
                print "Game Over!"
                break

def main():
    p=Game()
    p.play()

main()
